### PR TITLE
Support GCP GCR

### PR DESCRIPTION
**Pull request by hirosassa**
## Why

There's no GCR support on dockertags

## What

I implemented GCR support by communicating GCR API.

### CODE REVIEW

File: gcr.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+func fetchBearer(repo string, image string) (string, error) {
+	token, err := exec.Command("gcloud", "auth", "print-access-token").Output() // run gcloud command
+	if err != nil {
+		return "", err
+	}
+
+	url := constructGCRAuthURL(repo, image)
+	body, err := httpGet(url, "", &BasicAuthInfo{
+		Username: "_token",
+		Password: strings.TrimSpace(string(token)), // need to remove trailing new line character
```
Comment (bgpat) [inline gcr.go]: I would rather not use the external command to reduce dependencies. You can use [`FindDefaultCredentials`](https://pkg.go.dev/golang.org/x/oauth2/google?tab=doc#FindDefaultCredentials) instead.

```suggestion
	credential, err := google.FindDefaultCredentials(context.Background())
	if err != nil {
		return "", err
	}

	token, err := credential.TokenSource.Token()
	if err != nil {
		return "", err
	}

	url := constructGCRAuthURL(repo, image)
	body, err := httpGet(url, "", &BasicAuthInfo{
		Username: "_token",
		Password: token.AccessToken,
```

File: gcr.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		}
+	}
+	return tags
+}
+
+func retrieveFromGCR(repo string, image string) ([]string, error) {
+	bearer, err := fetchBearer(repo, image)
+	if err != nil {
+		return nil, err
+	}
```
Comment (bgpat) [inline gcr.go]: The public repository doesn't require an access token, so it should be continued even if it failed to get the token.

The following command is expected to succeed but now fails.

```console
$ /home/bgpat/bin/dockertags gcr.io/google_containers/hyperkube
exit status 1
```

File: gcr.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	"strings"
+)
+
+type gcrAuthResponse struct {
+	Token     string `json:"token"`
+	ExpiresIn int    `json:"expires_in"`
+}
+
+type gcrTagsResponse struct {
+	Manifest map[string]json.RawMessage `json:"manifest"`
```
Comment (bgpat) [inline gcr.go]: With nested struct, you can unmarshal at a time.

```suggestion
	Manifest map[string]json.gcrImageDetail `json:"manifest"`
```

File: gcr.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		return "", err
+	}
+
+	url := constructGCRAuthURL(repo, image)
+	body, err := httpGet(url, "", &BasicAuthInfo{
+		Username: "_token",
+		Password: strings.TrimSpace(string(token)), // need to remove trailing new line character
+	})
+
+	var resp gcrAuthResponse
```
Comment (bgpat) [inline gcr.go]: The API of GCR is compatible with Docker Hub.

```suggestion
	var resp DockerHubAuthResponse
```

File: gcr.go (hunk 3 above)
Comment (hirosassa) [reply gcr.go]: Thanks! I fixed it in e977236

File: gcr.go (hunk 4 above)
Comment (hirosassa) [reply gcr.go]: fixed in bab2dd0

File: gcr.go (hunk 2 above)
Comment (hirosassa) [reply gcr.go]: ignoring the error here in 872f215

File: gcr.go (hunk 1 above)
Comment (hirosassa) [reply gcr.go]: Thanks! 
I agree with this.
> I would rather not use the external command to reduce dependencies

fixed in 872f215

File: gcr.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+// for sorting setups
+type gcrImageDetail struct {
+	TimeUploadedMs string   `json:"timeUploadedMs"`
+	Tag            []string `json:"tag"`
+}
+
+type gcrImages []gcrImageDetail
+
+// ref: https://stackoverflow.com/questions/34037256/does-google-container-registry-support-docker-remote-api-v2/34046435#34046435
```
Comment (bgpat) [inline gcr.go]: delete this comment

Comment (hirosassa) [reply gcr.go]: OK, removed

**Review by bgpat — APPROVED** [review]
Thank you!! LGTM 👍

**Comment by hirosassa** [comment]
@bgpat Thank you for your a lot of valuable comment and suggestions!

**Pull request merged**